src/util/sss_krb5.c

```diff
@@ -416,9 +416,11 @@ krb5_error_code find_principal_in_keytab(krb5_context ctx,
     kerr = 0;
 
 done:
-    kerr_d = sss_krb5_free_keytab_entry_contents(ctx, &entry);
-    if (kerr_d != 0) {
-        DEBUG(SSSDBG_CRIT_FAILURE, "Failed to free keytab entry.\n");
+    if (principal_found) {
+        kerr_d = sss_krb5_free_keytab_entry_contents(ctx, &entry);
```

Comment:
I think strictly speaking this is not a double free since if the principal is not found entry should be memset to 0 and and the krb5_free_* calls should be able to handle that.

Reply:
It might be worth moving the first `memset(&entry, 0, ...)` to the top of the function for clarity.

Reply:
> I think strictly speaking this is not a double free since if the principal is not found entry should be memset to 0 and and the krb5_free_* calls should be able to handle that.

Doc - https://web.mit.edu/kerberos/krb5-current/doc/appdev/refs/api/krb5_free_keytab_entry_contents.html - doesn't promise this.

From my point of view, it is easier to be on a safe side.

Reply:
krb5 functions that free data all handle NULL, to match the behavior of `free()`.

Reply:
> krb5 functions that free data all handle NULL, to match the behavior of `free()`.

I am pretty sure they do. This is sane approach.
But is it stated officially in the docs somewhere?

Reply:
> It might be worth moving the first `memset(&entry, 0, ...)` to the top of the function for clarity.

Well, this is slight pessimization but really doesn't matter much. I do not have strong preference here... Ok, moved.

Reply:
Any project that *doesn't* define free(NULL) is dangerous and needs to be fixed.  No, it's not explicitly stated because we're reasonable people.  There's only one libkrb5.

Reply:
If it is not in the API docs then it is implementation specific and thus can't be relied upon.
This is the only reasonable approach of any user of 3rd party lib.